drivers/usb/gadget/function/f_scm.c

```diff
@@ -0,0 +1,574 @@
+{
+	struct usb_composite_dev *cdev;
+	struct f_scm *scm;
+	int id;
+	int ret;
+
+	cdev = c->cdev;
+	scm = func_to_scm(f);
+
+	id = usb_interface_id(c, f);
+	if (id < 0)
+		return -ENODEV;
+
+	scm_intf.bInterfaceNumber = id;
+
+	id = usb_string_id(cdev);
+	if (id < 0)
+		return -ENODEV;
+
+	scm_string_defs[0].id = id;
+	scm_intf.iInterface = id;
+
+	/* Set up the bulk and command endpoints */
+	scm->bulk_in = usb_ep_autoconfig(cdev->gadget, &fs_scm_out_desc);
+	if (!scm->bulk_in) {
+		ERROR(cdev, "%s: can't autoconfigure bulk source on %s\n",
+			f->name, cdev->gadget->name);
+		return -ENODEV;
+
+	}
+
+	scm->bulk_out = usb_ep_autoconfig(cdev->gadget, &fs_scm_in_desc);
+	if (!scm->bulk_out) {
+		ERROR(cdev, "%s: can't autoconfigure bulk sink on %s\n",
+			f->name, cdev->gadget->name);
+		return -ENODEV;
+
+	}
+
+	scm->cmd_out = usb_ep_autoconfig(cdev->gadget, &fs_scm_cmd_in_desc);
+	if (!scm->cmd_out) {
+		ERROR(cdev,
+			"%s: can't autoconfigure control source on %s\n",
+			f->name, cdev->gadget->name);
+		return -ENODEV;
+	}
+
+	scm->cmd_in = usb_ep_autoconfig(cdev->gadget,
+		&fs_scm_cmd_out_desc);
+	if (!scm->cmd_in) {
+		ERROR(cdev, "%s: can't autoconfigure control sink on %s\n",
+		f->name, cdev->gadget->name);
+		return -ENODEV;
+	}
+
+	/* support high speed hardware */
+	hs_scm_out_desc.bEndpointAddress =
+		fs_scm_out_desc.bEndpointAddress;
+	hs_scm_in_desc.bEndpointAddress =
+		fs_scm_in_desc.bEndpointAddress;
+	hs_scm_cmd_out_desc.bEndpointAddress =
+		fs_scm_cmd_out_desc.bEndpointAddress;
+	hs_scm_cmd_in_desc.bEndpointAddress =
+		fs_scm_cmd_in_desc.bEndpointAddress;
+
+	/* support super speed hardware */
+	ss_scm_out_desc.bEndpointAddress =
+		fs_scm_out_desc.bEndpointAddress;
+	ss_scm_in_desc.bEndpointAddress =
+		fs_scm_in_desc.bEndpointAddress;
+	ss_scm_cmd_out_desc.bEndpointAddress =
+		fs_scm_cmd_out_desc.bEndpointAddress;
+	ss_scm_cmd_in_desc.bEndpointAddress =
+		fs_scm_cmd_in_desc.bEndpointAddress;
+
+	/* Copy the descriptors to the function */
+	ret = usb_assign_descriptors(f, scm_fs_descs, scm_hs_descs,
+			scm_ss_descs, NULL);
+	if (ret)
+		goto fail;
+
+	DBG(cdev, "SCM bind complete at %s speed\n",
+		gadget_is_superspeed(c->cdev->gadget) ? "super" :
+		gadget_is_dualspeed(c->cdev->gadget) ? "dual" : "full");
+fail:
+	return ret;
+}
+
+static void scm_free_func(struct usb_function *f)
+{
+	struct f_scm_opts *opts;
+
+	opts = container_of(f->fi, struct f_scm_opts, func_inst);
+
+	mutex_lock(&opts->lock);
+	opts->refcnt--;
+	mutex_unlock(&opts->lock);
+
+	usb_free_all_descriptors(f);
+	kfree(func_to_scm(f));
+}
+
+static int enable_endpoint(struct usb_composite_dev *cdev, struct f_scm *scm,
+	struct usb_ep *ep)
+{
+	int result;
+
+	result = config_ep_by_speed(cdev->gadget, &(scm->function), ep);
+	if (result)
+		goto out;
+
+	result = usb_ep_enable(ep);
+	if (result < 0)
+		goto out;
+	ep->driver_data = scm;
+	result = 0;
+out:
+	return result;
+}
+
+/**
+ * @todo add error out that disables endpoint when fail
```

Comment:
Need to address this now.  Right now the function is just ignoring (actually logging, but proceeding) on errors for the first 3 endpoints.  It needs to disable the already enabled endpoints and return an error code instead.